docs/docs/Best_Practices/Metadata_and_Message_Limits.md

```diff
@@ -0,0 +1,28 @@
+# Metadata and Message Limits
+Operations that exceed their internal metadata and/or message limits will be terminated.
+
+## Maximum size for metadata
+The maximum size for metadata is 32 megabytes (`32MB`).
+
+The following fields within [User Visible Errors](/Best_Practices/User_Visible_Errors.md) will all count towards used internal metadata:
+1. `message`
+2. `action`
+3. `output`
```

Comment:
Are user-visible errors the only thing that counts as "metadata"?  Will readers know what is meant here by "metadata", or should we have a (possibly informal) definition of that term, too?

Reply:
I find this section a little hard to communicate to the plugin author, so any suggestions would be great.

Originally, I wrote out a section describing what "internal metadata" means, but I ended up leaking too much implementation detail into the docs, so I decided to remove it.

In a nutshell, what I'm trying to tell the plugin author is that the plugin will communicate with the libs server and platform client in it's lifecycle and that the plugin author has control over the metadata and message size through the arguments passed into the libs and platform endpoints. It was particularly hard address this in the docs without spewing out information about gRPC services, HTTP, and docker.

Reply:
would it make sense to get like a code example of like say, if some user error was defined a certain way, what part of that message is part of the metadata and therefore part of the limit?

Otherwise idk if there's really any words that could describe this better

Reply:
One thing I noticed is that it seems like the limit for a message is much smaller than the limit for an metadata object?   That doesn't make a ton of sense to me, since every metadata object will appear in at least one message (e.g. a snapshot object must always be returned from a plugin operation in a message, right?).

So, it seems, practically, that the 32MB limit on object size is not important here, since the 128KiB limit for messages will be exceeded first.

Anyhow, given that, here's my attempt to reword this:




OVERVIEW

There are limits on how much data can be passed back and forth between the engine and the plugin at one time.  The five places where that comes into play are:

- Inputs sent from the engine to the plugin, as arguments to plugin operations(link). For example, the schema-defined `repository` object that is provided to most plugin operations.

- Outputs sent back from the plugin to the engine, as the return values of plugin operations. For example, the schema-defined `snapshot` object that is returned from the plugin's `post_snapshot` method.

- Inputs sent from the plugin to the engine, as arguments to Platform Library functions(link).  For example, the string message passed to `logger.debug`.

- Outputs sent back from the engine to the plugin, as return values from Platform Library functions. For example, the stdout resulting from a call to `libs.run_bash`.

- Exception objects thrown by plugin code.


MESSAGE LIMITS

For any of these five cases, the sum total size of all the information passed along must be less than 128Kb.

The actual size of this information at runtime is somewhat dependent on how the Python interpreter chooses to represent the information, so it's not always possible to know ahead of time exactly what the size will be.

However, here are some examples of where problems are most likely:

- Using `run_bash` to print the entire contents of a large file ot stdout.
- Using a single `logger` command with many pages of output.
- Throwing an exception object that includes large objects inside of it.

Reply:
So I just realized after reading your comment that the maximum sizes are swapped for metadata and messages (I was confused as well). The maximum size for metadata is 128KiB whereas the maximum size for messages is 32mb. This makes a lot more sense so thanks for catching that!